text/trace/0234-payload-collection.md

```diff
@@ -0,0 +1,363 @@
+
+Ultimately, it could be beneficial if OpenTelemetry instrumentation SDKs will have OOTB optional payload collection as well, though this OTEP does NOT propose that, or takes into account that it will be added in the future.
+
+## Explanation
+
+While payload data can just be a binary buffer, most of the times it has a defined structure.
+For modern applications and APIs, the payload data is usually encoded as one of:
+
+- JSON
+- YAML
+- Protobuf
+- Avro
+- Plain strings
+- Blob
+
+Whenever the payload data has some meaningful structure (everything except a blob), that structure will be reflected in the semantic attribute of the relevant span. That way, processors downstream can easily access and manipulate the payload data.
+
+## Internal details
+
+### Semantic conventions
+
+We propose specifying a consistent naming for payload content attributes, and its
+related metadata. These conventions could help processors, backends and users to
+handle this kind of data.
+
+In this OTEP we only give ideas and recommendations for it, though
+a further discussion will be required for choosing the ideal naming when
+writing the actual specifications.
+
+The conventions we propose are added as a postfix to the base attribute name
+(for example, `http.request.body`):
+
+- **Data attribute**: `<attribute>.content`. Holds the decoded content of the payload data. Alternatives: `.data`, `.payload`.
+- **Size attribute**: `<attribute>.size`. Holds the number of bytes of the encoded payload data. Alternatives: `.length`, `.bytes`.
+- **Encoding attribute**: `<attribute>.encoding`. Holds the original attribute encoding type.
+  Predefined values should be declared (though users may decide using custom values as well).
+  For example - `json`, `protobuf`, `avro`, `utf-8`.
+
+Adding each attribute should not be dependent on the others in any way.
+
+### API
+
+We propose specifying APIs the will abstract the handling and capturing of
+payload data, to allow customizing this functionality and evolving it over time.
+
+This functionality includes:
+
+- Adding attributes with enforced semantics
+- Enforcing limits - e.g. shortening long strings
+- Applying central configuration over the general functionality
+- Parsing raw payload buffers - converting payload data from bytes to OTel attributes, for
+  supported encodings, in an extendable fashion.
+- Aggregating multiple data chunks (to assist with handling asynchronous buffers)
+
+The APIs given here are only a draft to describe general characteristics.
+Like in the semantics, a further discussion will be required before creating actual specifications.
+
+#### `Payload` class
+
+We propose adding a class that will provide the related functionality for handling payload data.
+Both traces and logs could use instances of this class, which will assist sharing related functionality.
```

Comment:
Since this was added in real tracing products, can you link to the APIs/SDKs they offered?

Reply:
It's a bit different, since they just offered this instrumentation OOTB, so the API for the payload was hidden from the developer. See for example [Epsagon's instrumentation for FastAPI](https://github.com/epsagon/epsagon-python/blob/master/epsagon/wrappers/fastapi.py) (The exact structure of the data is [seen here](https://github.com/epsagon/epsagon-python/blob/master/epsagon/runners/fastapi.py#L78))